lib/relooper/src/graph/domtree.rs

```diff
@@ -0,0 +1,196 @@
+        self.reachable(dominator)
+    }
+
+    pub fn is_stdom(&self, dominator: &T, dominated: &T) -> bool {
+        dominator != dominated && self.is_reachable(dominator, dominated)
+    }
+
+    pub fn imm_dominated(&self, n: &T) -> Option<&T> {
+        for (from, to) in self.edges() {
+            if to.contains(n) {
+                return Some(from);
+            }
+        }
+        None
+    }
+
+    pub fn is_idom(&self, dominator: &T, dominated: &T) -> bool {
+        self.children(dominator).contains(dominated)
+    }
+
+    pub fn level(&self, node: &T) -> usize {
+        *self
+            .levels
+            .get(node)
+            .expect("node should be present in domination tree in order to acquire its level")
+    }
+
+    pub fn levels(&self) -> &HashMap<T, usize> {
+        &self.levels
+    }
+
+    pub fn by_level(&self) -> BTreeMap<usize, HashSet<T>> {
+        let mut res: BTreeMap<usize, HashSet<T>> = Default::default();
+
+        for (x, &level) in &self.levels {
+            res.entry(level).or_default().insert(*x);
+        }
+
+        res
+    }
+
+    pub fn max_level(&self) -> usize {
+        *self.levels.iter().map(|(_, level)| level).max().unwrap()
+    }
+
+    fn update_dominators(
+        cfg: &Cfg<T>,
+        node_ordering: &NodeOrdering<T>,
+        cur_id: T,
+        origin: T,
+        result: &mut HashMap<T, T>,
+    ) {
+        let mut reachable_set = HashSet::<T>::default();
+        for &node in node_ordering.sequence() {
+            reachable_set.insert(node);
+        }
+
+        let reached = Dfs::start_from(&origin, |&n| {
+            let mut ch = cfg.children(&n);
+            ch.remove(&cur_id);
+            ch
+        });
+        for id in reached {
+            reachable_set.remove(id);
+        }
+        reachable_set.remove(&cur_id);
+        for id in reachable_set {
+            result.insert(id, cur_id);
+        }
+    }
+
+    pub fn domination_map(cfg: &Cfg<T>, node_ordering: &NodeOrdering<T>) -> HashMap<T, T> /* map points from node id to id of its dominator */
+    {
+        let mut result = HashMap::<T, T>::new();
+        let mut bfs = VecDeque::<T>::new();
+        let mut visited = HashSet::<T>::new();
+        for &n in node_ordering.sequence() {
+            result.insert(n, cfg.entry);
+        }
+        bfs.push_back(cfg.entry); // should be next. upd: i dont think so
+        visited.insert(cfg.entry);
+        loop {
+            if bfs.is_empty() {
+                break;
+            }
+            let &cur_id = bfs.front().unwrap();
+            visited.insert(cur_id);
+            bfs.pop_front().unwrap();
+            Self::update_dominators(cfg, node_ordering, cur_id, cfg.entry, &mut result);
+            for &id in cfg.children(&cur_id) {
+                if !visited.contains(&id) {
+                    bfs.push_back(id);
+                }
+            }
+        }
+        result
+    }
+
+    //todo remove with new dominance building
```

Comment:
Can this be removed now? Am I understanding correctly that the new DJ building stuff replaces the old domination tree building?

Reply:
we are still using old domination tree building for that: its result is passed to 'DomTree::from_edges' as 'domination_map'

Reply:
Is the new domination tree building finished in another branch? Or was it not fully implemented?

Reply:
as we discussed yesterday, that bit of work will be done by Viktar in another pr on top of current changes